### PR TITLE
storage/etcd3: factor tests to accept `storage.Interface`

### PR DESCRIPTION
**Pull request by stevekuznetsov**
storage/etcd3: factor tests to accept `storage.Interface`

This commit simply factors the test functionality into functions that
accept `storage.Interface`.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

storage/etcd3: move reusable tests to importable files

These can't live in `_test.go` files, or others won't be able to re-use
them as you can't import from test files.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

/kind cleanup
/sig api-machinery

```release-note
NONE
```

```docs

```

/assign @wojtek-t 
/cc @liggitt @lavalamp

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
Part of https://github.com/kubernetes/kubernetes/issues/109831

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_tests.go [hunk 1]
```diff
@@ -0,0 +1,742 @@
...
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"reflect"
+	"strconv"
+	"sync"
+	"testing"
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_tests.go:27]: anything that imports `testing` either needs to be in a `_test.go` file or in a dedicated `testing` subpackage... putting this here means that the `testing` package gets linked into production binaries

Comment (stevekuznetsov) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_tests.go:27]: ACK - @wojtek-t looks like there's no way around a new package for this.

**Comment by stevekuznetsov** [comment]
@liggitt @wojtek-t updated to use the `storage/testing` package. We can revert all of the exported helper methods when the rest of the tests are moved over.

**Comment by stevekuznetsov** [comment]
Flake on `TestDSCUpdatesPodLabelAfterDedupCurHistories` 
/retest

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_tests.go (hunk 1 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_tests.go]: OK - we'll figure out the 0% coverage aspect of that test later.

**Comment by wojtek-t** [comment]
Thanks - this looks great.

/lgtm
/approve

**Pull request merged**